pennylane/transforms/qcut.py

```diff
@@ -210,3 +207,80 @@ def tape_to_graph(tape: QuantumTape) -> MultiDiGraph:
             order += 1
 
     return graph
+
+
+def fragment_graph(graph: MultiDiGraph) -> Tuple[Tuple[MultiDiGraph], MultiDiGraph]:
+    """
+    Fragments a graph into a collection of subgraphs as well as returning
+    the communication/`quotient <https://en.wikipedia.org/wiki/Quotient_graph>`__
+    graph. Each node of the commnication graph represents a fragment and the edges
+    denote the flow of qubits between fragments.
+
+    Args:
+        graph (MultiDiGraph): directed multigraph containing measure and prepare
+            nodes at cut locations
+
+    Returns:
+        Tuple[Tuple[MultiDiGraph], MultiDiGraph]: the subgraphs of the cut graph
+        and the communication graph.
+
+    **Example**
+
+    Consider the following circuit with manually-placed wire cuts:
+
+    .. code-block:: python
+
+        wire_cut_0 = qml.WireCut(wires=0)
+        wire_cut_1 = qml.WireCut(wires=1)
+        multi_wire_cut = qml.WireCut(wires=[0, 1])
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(0.4, wires=0)
+            qml.apply(wire_cut_0)
+            qml.RY(0.5, wires=0)
+            qml.apply(wire_cut_1)
+            qml.CNOT(wires=[0, 1])
+            qml.apply(multi_wire_cut)
+            qml.RZ(0.6, wires=1)
+            qml.expval(qml.PauliZ(0))
+
+    We can find the corresponding graph, remove all the wire cut nodes, and
+    find the subgraphs and communication graph by using:
+
+    >>> graph = qml.transforms.tape_to_graph(tape)
+    >>> qml.transforms.replace_wire_cut_nodes(graph)
+    >>> qml.transforms.fragment_graph(graph)
+    ((<networkx.classes.multidigraph.MultiDiGraph object at 0x7fb3b2311940>,
+      <networkx.classes.multidigraph.MultiDiGraph object at 0x7fb3b2311c10>,
+      <networkx.classes.multidigraph.MultiDiGraph object at 0x7fb3b23e2820>,
+      <networkx.classes.multidigraph.MultiDiGraph object at 0x7fb3b23e27f0>),
+     <networkx.classes.multidigraph.MultiDiGraph object at 0x7fb3b23e26a0>)
+    """
+
+    graph_copy = copy.deepcopy(graph)
+
+    edges = list(graph_copy.edges)
+    cut_edges = []
+
+    for node1, node2, wire in edges:
```

Comment:
```suggestion
    graph_copy = graph.copy()

    cut_edges = []

    for node1, node2, wire in graph.edges:
```
Why do we need to do a deep copy? I saw @zeyueN you mentioned potentially deep copying, did you have an intuition on whether a shallow copy would be insufficient?

Reply:
Oh good point, it's just me being not familiar with how nx graphs handles their `__copy__()`. But based on [their doc](https://networkx.org/documentation/stable/reference/classes/generated/networkx.Graph.copy.html), it seems a shallow copy should suffice.